zaza/openstack/charm_tests/heat/tests.py

```diff
@@ -0,0 +1,355 @@
+        unit = zaza.model.get_unit_from_name(unit.entity_id)
+        assert unit.workload_status == "active"
+
+    def test_400_heat_resource_types_list(self):
+        """Check default resource list behavior and confirm functionality."""
+        logging.info('Checking default heat resource list...')
+        try:
+            types = list(self.heat_client.resource_types.list())
+            if type(types) is list:
+                logging.info('Resource type list check is ok.')
+            else:
+                msg = 'Resource type list is not a list!'
+                logging.error('{}'.format(msg))
+                raise
+            if len(types) > 0:
+                logging.info('Resource type list is populated '
+                             '({}, ok).'.format(len(types)))
+            else:
+                msg = 'Resource type list length is zero!'
+                logging.error(msg)
+                raise
+        except Exception as e:
+            msg = 'Resource type list failed: {}'.format(e)
+            logging.error(msg)
+            raise
+
+    def test_402_heat_stack_list(self):
+        """Check default heat stack list behavior, confirm functionality."""
+        logging.info('Checking default heat stack list...')
+        try:
+            stacks = list(self.heat_client.stacks.list())
+            if type(stacks) is list:
+                logging.info("Stack list check is ok.")
+            else:
+                msg = 'Stack list returned something other than a list.'
+                logging.error(msg)
+                raise
+        except Exception as e:
+            msg = 'Heat stack list failed: {}'.format(e)
+            logging.error(msg)
+            raise
+
+    def test_410_heat_stack_create_delete(self):
+        """Create stack, confirm nova compute resource, delete stack."""
+        logging.info('Creating, deleting heat stack (compute)...')
+        self._image_create()
+        self._keypair_create()
+        self._stack_create()
+        self._stack_resource_compute()
+        self._stack_delete()
+        self._image_delete()
+        self._keypair_delete()
+
+    def test_500_auth_encryption_key_same_on_units(self):
+        """Test the auth_encryption_key in heat.conf is same on all units."""
+        logging.info("Checking the 'auth_encryption_key' is the same on "
+                     "all units.")
+        output, ret = self._run_arbitrary(
+            "--application heat "
+            "--format json "
+            "grep auth_encryption_key /etc/heat/heat.conf")
+        if ret:
+            msg = "juju run returned error: ({}) -> {}".format(ret, output)
+            logging.error("Error: {}".format(msg))
+        output = json.loads(output)
+        keys = {}
+        for r in output:
+            k = r['Stdout'].split('=')[1].strip()
+            keys[r['UnitId']] = k
+        # see if keys are different.
+        ks = list(keys.values())
+        if any(((k != ks[0]) for k in ks[1:])):
```

Comment:
A quicker/easier way to do this is to use a set:

    ks = set(keys.values())
    self.assertEqual(len(ks), 1, "'auth_enc....: {}".format(", ".join(ks)))

Reply:
@ajkavanagh Trying to understand what the right code should be here. If it's OK with you, I will treat this as Optional for now (or at least put it at the bottom of my list).

Thank you.

Reply:
Okay, which is easier to understand:

```python
ks = list(keys.values())
if any(((k != ks[0]) for k in ks[1:])):
```

or

```python
ks = set(keys.values())
if len(ks) != 1:
```

because they do exactly the same thing.   It's just that a `set()` automatically only has unique members.  i.e. if the keys are different then there will be more than 1 key in the set.

Reply:
Thanks @ajkavanagh, this is much simpler. Full disclaimer: I was using the same code that was in the original Amulet testing.